doc/acmerge.1
.\" $Id: acmerge.1,v 1.4 2024/07/12 22:56:16 tom Exp $
.TH ACMERGE 1 2024-07-12 MiscTools "User commands"
.ie \n(.g \{\
.ds `` \(lq
.ds '' \(rq
.\}
.el \{\
.ie t .ds `` ``
.el   .ds `` ""
.ie t .ds '' ''
.el   .ds '' ""
.\}
.hy 0
.SH NAME
acmerge \-
merge a split aclocal.m4
.SH SYNOPSIS
.B acmerge
[
.I filename
]
.
.SH DESCRIPTION
Merge parts for a split-up autoconf source file (usually
aclocal.m4) from a parent file which was processed by
\fBacsplit\fP.
.PP
The command parameter tells \fBacmerge\fP where to write merged autoconf
source (default: \*(``aclocal.m4\*('').
.
.SH OPTIONS
.TP 5
.B \-n
do not create a comment-line with dashes between macros in the resulting file.
.TP 5
.B \-v
verbose
.
.
.SH BUGS
.
Probably.
.
.SH EXAMPLES
.
\fBAcmerge\fR is used to construct aclocal.m4 from the macros which are
stored in a repository.
.
.SH SEE ALSO
acsplit(1)


CHANGES
2025-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/sue/sue.c: workaround for broken NetBSD header

	* td_getopt.h: NetBSD lint

	* package/debian/changelog, package/misc_tools.spec: bump

2025-09-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* COPYING.asc: sign

	* package/misc_tools.spec: SPDX equivalent

	* COPYING: consolidate license

	* install-bin.in:
	amend $INSTALL in case install-sh is used from a subdirectory

	* install-bin.in: shellcheck warnings

	* config_h.in, doc/minstall.in, package/debian/postinst,
	  src/newpath/makefile.wnt, configure.in:
	trim-blanks

	* src/unmap/chrcount.c, src/unmap/map.c, src/unmap/map_s.c,
	  src/unmap/unmap.c, src/unmap/unmap_s.c, src/width/width.c,
	  src/splitit/splitit.c, src/perror/perror.c, src/slowcat/slowcat.c,
	  src/sue/sue.c, src/tisplit/timerge.c, src/tisplit/tisplit.c,
	  src/acsplit/acmerge.c, src/count_files/count_files.c, src/cpd/cpd.c,
	  src/newest/newest.c, src/newpath/newpath.c, src/acsplit/acsplit.c,
	  td_getline.h:
	s/0/NULL/

	* doc/newpath.1: bump

	* doc/newpath.1: grammar

	* doc/cpd.1: bump

	* doc/cpd.1: grammar

	* doc/chrcount.1: bump

	* doc/chrcount.1: grammar

	* doc/acsplit.1: bump

	* doc/acsplit.1: grammar

	* package/debian/copyright: bump

	* configure: regen

	* package/debian/changelog, package/misc_tools.spec: bump

2025-08-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: update to 2024-07-27

2024-07-15  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-bin.in: source parameter may have $DESTDIR prefix

	* src/sue/use-usual:
	use $DestDir rather than $DESTDIR, and use the install program
	to install the prepared wrapper rather than fixing after install.

	* package/misc_tools.spec: use --with-execdir option

	* src/sue/use-hacks: quoting per shellcheck warnings
	do not exit on the first error, to work with rpmbuild

	* package/debian/changelog: bump

2024-07-13  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/newpath/newpath.c:
	a relative link for newpath, i.e., making /usr/bin/newpath a symbolic link
	to ../libexec/misctools/newpath did not work -- fix this by finding the
	program on the path, and using realpath to resolve links in that.

	* install-bin.in: corrected link source

	* package/debian/postinst: using libexec now, rather than lib

	* src/sue/use-hacks:
	s/DESTDIR/DestDir/ to avoid conflict with $DESTDIR used in package builds

	* configure: regen

	* src/sue/use-hacks: exit on the first error

	* install-bin.in: add special case for src/sue

	* configure.in: add install_bin symbol, for special case in src/sue

	* src/sue/makefile.in:
	bin-directory depends upon the --with-execdir option

	* configure: regen

	* install-bin.in: account for DESTDIR

	* configure.in: improve message-result for logging

	* package/debian/postinst:
	symbolic link is now done via configure option --with-execdir

	* package/debian/rules: use --with-execdir option

	* configure: regen

	* aclocal.m4: add CF_PROG_LN_S

	* install-bin.in: RCS_BASE

	* package/misc_tools.spec:
	simplify by adding a configure option --with-execdir

	* configure.in, src/acsplit/makefile.in, src/count_files/makefile.in,
	  src/cpd/makefile.in, src/hex/makefile.in, src/isatty/makefile.in,
	  src/newest/makefile.in, src/newpath/makefile.in,
	  src/perror/makefile.in, src/realpath/makefile.in,
	  src/slowcat/makefile.in, src/splitit/makefile.in, src/sue/makefile.in,
	  src/tisplit/makefile.in, src/unmap/makefile.in, src/width/makefile.in:
	added --with-execdir option to simplify installing the tools into libexecdir
	while making a symbolic link to newpath

	* makefile.in: provide for cleanup of install-bin script

	* package/misc_tools.spec:
	build all programs, since there is no naming conflict now

	* doc/minstall.in: update to match unquoted section number in .TH line

	* package/debian/changelog, package/misc_tools.spec: bump

2024-07-12  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/misc_tools.spec:
	add relative symlink for newpath to the %files list

	* doc/newpath.1:
	explain how to reduce environment conflicts with this set of utilities

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	quoting

	* doc/newpath.1: identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/newpath.1, doc/perror.1, doc/realpath.1, doc/slowcat.1,
	  doc/splitit.1, doc/sue.1, doc/timerge.1, doc/tisplit.1, doc/unmap.1,
	  doc/width.1:
	check-manpage

	* package/debian/copyright: bump

	* configure: regen

	* COPYING.asc: sign

	* COPYING, package/debian/changelog, package/misc_tools.spec: bump

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: update to 2024-05-27

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023/12/27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023/12/2

2023-12-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-04-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* doc/minstall.in: reduce filename-conflicts by installing ".1mt" files

	* configure: regen


src/acsplit/acmerge.c
/*
 * $Id: acmerge.c,v 1.13 2025/09/10 20:38:30 tom Exp $
 *
 * Title:	acmerge.c - merge a split aclocal.m4
 * Author:	T.E.Dickey
 * Created:	28 Aug 1997
 * Function:	Merges parts for a split-up autoconf macro file (usually
 *		aclocal.m4) from a parent file which was processed by acsplit,
 *		and the macros in a subdirectory.
 */
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

#include <td_getline.h>
#include <td_getopt.h>

#define isname(c) ((isalnum((unsigned char)c) || (c) == '_'))

#define VERBOSE(n) if(verbose > n) printf

static const char *target = "AcSplit";

static int do_dash = 1;
static int verbose = 0;

static void
failed(char *msg)
{
    perror(msg);
    exit(EXIT_FAILURE);
}

static char *
skip_blanks(char *s)
{
    while (isspace((unsigned char) *s))
	s++;
    return s;
}

static int
is_comment(char *s)
{
    if (!strncmp(s, "dnl", 3)) {
	s += 3;
	if (isspace((unsigned char) *s) || ispunct((unsigned char) *s))
	    return 1;
    }
    return 0;
}

static char *
skip_comment(char *s)
{
    if (!strncmp(s, "dnl", 3)) {
	s += 3;
	if (isspace((unsigned char) *s))
	    return skip_blanks(s);
    }
    return NULL;
}

static int
is_dashes(char *line)
{
    line = skip_blanks(line);
    line = skip_comment(line);
    if (line != NULL && strlen(line) >= 8) {
	return (!strncmp(line, "--------", 8));
    }
    return 0;
}

static void
dashes(FILE *ofp)
{
    if (do_dash) {
	int n;

	fputs("dnl ", ofp);
	for (n = 0; n < 75; n++)
	    fputc('-', ofp);
	fputc('\n', ofp);
    }
}

static void
append(char *name, FILE *ofp)
{
    size_t have = 0;
    char *temp = malloc(have = BUFSIZ);
    char *t;
    FILE *ifp;
    int found = 0;
    int count = 0;
    char *next = skip_blanks(name);

    sprintf(temp, "%s/", target);
    t = temp + strlen(temp);
    while (isname(*next)) {
	*t++ = *next++;
	found = 1;
    }
    *t = 0;

    VERBOSE(0) ("appending %s\n", temp);
    if ((ifp = fopen(temp, "r")) == NULL) {
	if (!strncmp(name, "AC_", 3) || !strncmp(name, "AM_", 3))
	    return;
	failed(temp);
    }

    if (found)
	dashes(ofp);
    else
	fputc('\n', ofp);
    while (getline(&temp, &have, ifp) >= 0) {
	fputs(temp, ofp);
	++count;
    }
    VERBOSE(0) ("...appended %d lines\n", count);
    fclose(ifp);
    free(temp);
}

static void
acmerge(const char *path)
{
    FILE *hdr;
    FILE *ofp;
    char name[BUFSIZ];
    char temp[BUFSIZ];
    char *bfr = NULL;
    size_t have = 0;

    sprintf(name, "%s.in", path);
    if ((hdr = fopen(name, "r")) == NULL)
	failed(name);

    sprintf(temp, "%s.out", target);
    remove(temp);
    if ((ofp = fopen(temp, "w")) == NULL)
	failed(temp);

    while (getline(&bfr, &have, hdr) >= 0) {
	VERBOSE(0) ("read %s", bfr);
	if (is_dashes(bfr)) {
	    dashes(ofp);
	} else if (is_comment(bfr)) {
	    fputs(bfr, ofp);
	} else {
	    append(bfr, ofp);
	}
    }
    fclose(hdr);
    fclose(ofp);
    rename(temp, path);
    free(bfr);
}

static void
usage(void)
{
    fprintf(stderr, "Usage: acmerge [-n] [-v] [aclocal.m4]\n");
    exit(EXIT_FAILURE);
}

int
main(int argc, char *argv[])
{
    int c;

    while ((c = getopt(argc, argv, "nv")) != -1) {
	switch (c) {
	case 'n':
	    do_dash = 0;
	    break;
	case 'v':
	    ++verbose;
	    break;
	default:
	    usage();
	    break;
	}
    }
    if (optind < argc) {
	while (optind < argc) {
	    acmerge(argv[optind++]);
	}
    } else {
	acmerge("aclocal.m4");
    }

    return EXIT_SUCCESS;
}
